### DISCUSSION TITLE
vsgconv - Setting boolean options to false?

### DISCUSSION THREAD

**Discussion by geefr** [vsg-dev/VulkanSceneGraph#617] [General]
Hopefully an easy question, or I'm missing something obvious, but I don't seem to be able to turn boolean options like two_sided _off_ when using vsgconv (I'm trying to debug why I'm getting lighting from behind an object, whether it's this or my vr integration stuff messing with the lighting again).

Running `vsgconv foo.glb foo.vsgt --crease_angle 10` or similar for numeric arguments seems to work fine

Running `vsgconv foo.glb foo.vsgt --two_sided 0` doesn't - Gets confused over what's an option vs a path

Looking in utils/CommandLine.h we see this - I can't see what command line argument I should provide to turn the boolean flag off. In this case two_sided is I think on by default because of how my materials were exported to gltf from blender. I'd prefer to just toggle it back off on the command line for now.

```
 template<typename T>
        bool readAndAssign(const std::string& match, Options* options)
        {
            if constexpr (std::is_same_v<T, void>)
            {
                if (options && read(std::string("--") + match))
                {
                    options->setValue(match, true);
                    return true;
                }
            }
            else // numeric args
```

**Comment by robertosfield** [comment]
The assimp reader is not looking for any other parameters when reading --two-sided. not the void in the code vsgXchange/src/assimp/assimp.cpp:

```
bool assimp::readOptions(vsg::Options& options, vsg::CommandLine& arguments) const
{
    bool result = arguments.readAndAssign<void>(assimp::generate_smooth_normals, &options);
    result = arguments.readAndAssign<void>(assimp::generate_sharp_normals, &options) || result;
    result = arguments.readAndAssign<float>(assimp::crease_angle, &options) || result;
    result = arguments.readAndAssign<void>(assimp::two_sided, &options) || result;
    return result;
}
```
In this case it's treating the presence of --two-sided as true.  The default will be off, so forcing it one will be the normal thing that users will likely want.  If we need more command line control over it we'll need to add another option check.

The code for deciding whether two sided is used is when setting up the shader composition is:

```
    if ((options->getValue(assimp::two_sided, convertedMaterial.two_sided) || (material->Get(AI_MATKEY_TWOSIDED, convertedMaterial.two_sided) == AI_SUCCESS)) && convertedMaterial.two_sided)
    {
        defines.insert("VSG_TWO_SIDED_LIGHTING");
    }
```

Which looks like if either two_sided is passed in via the Options (which in turn can be set from the command line) or from the material settings provided by the assimp library when reading the file.  I presume in your case the actual model is specifying two sided lighting which is why it's not off by default.

To override this I think it should be possible to do:

```
   options->setValue("two_sided", false);
   auto model = vsg::read_cast<vsg::Node>("my_model", options);
```

Being able to do this from the command line would be nice, but like the convenience of passing a single --option_name over --option_name value for things that are normally off and adding the --option will enable it.  Perhaps having the value be optional i.e.  checked for --option_name true,  --option_name false first then  --option_name with implicit true.

I'm open to suggestions.

**Comment by geefr** [comment]
Ah that makes sense. If it's possible, being able to `--option=false` would be nice there. Not a huge blocker for me since my models are quite simple, and I really should be fixing the problem at the source.

It does turn out the default for backface culling in Blender is off - So once exported to gltf, if you forgot to turn on backface culling for each material it'll translate to two_sided in the vsg conversion, and look somewhat different than the original Blender scene. I assume that's blender doing shadows/fancier shaders, so even without culling the light can't pass through an object.

**Comment by robertosfield** [comment]
I have [added](4efc9262cc3a73d49db9abcd44529abf6257f9b5) specialization of the vsg::CommandLine::read(..) methods for the bool case that enable the following usage:

```
# all set toggle to false
myapp --toggle false
myapp --toggle False
myapp --toggle FALSE
myapp --toggle 0

# all set toggle to true
myapp --toggle true
myapp --toggle True
myapp --toggle  TRUE
myapp --toggle 1
myapp --toggle

```
I have also changed vsgXhcnage::assimp to use bool rather than void for command line parsing of bool paramters as this wasn't possible before as I wanted the --toggle (without parameter usage to be the default usage) for enabling a feature.

These changes are merged with VulkanSceneGraph and vsgXchange master and respective 1.0 branches.
